File: src/data_rover/core/navigation/schema.py
"""Navigation definitions: walk chains of relationships from a filtered
start set, or combine navigation results with set operations.

Format contract (Stage 1 of the navigation/tables/diagrams mega-plan):

- The document is VERSIONED (`schema_version`) and branch-READY: every
  relationship step carries a `children` slot so branching can be added later
  without migrating saved artifacts — but v2 validation rejects non-empty
  `children` (linear chains only) and caps a definition's total step items at
  MAX_STEPS.
- Filters reuse the advanced-search criterion vocabulary
  (`core.search.criteria`) verbatim, so a navigation filter and a search
  criterion are the same wire object.
- CHAIN CONVENTION: an evaluated chain includes its start element at index 0
  (a path with N *relationship or property* steps yields chains of length N+1 — a
  `FilterStep` prunes the frontier in place and adds no column).
  `Operand.step_index` addresses that tuple directly: 0 = start elements,
  k = elements after chain-advancing step k, None = terminal step.
- Set operations act on ELEMENT SETS (the elements a navigation reaches at
  one step), never on chain tuples; `difference` is a left fold over the
  operand list.

Definitions are stored as `kind="navigation"` project artifacts; `Operand.ref`
holds such an artifact's id and is inlined by `core.navigation.resolve` before
evaluation (the evaluator itself never sees a ref).
"""

from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, Field, TypeAdapter, model_validator

from data_rover.core.search.criteria import Criterion

SCHEMA_VERSION = 3
#: hard ceiling on chain length; also the recursion depth of the evaluator.
MAX_STEPS = 10


class Scope(BaseModel):
    """An element filter: any of `types` (subtype-inclusive; empty = every
    type) AND all of `criteria`."""

    kind: Literal["scope"] = "scope"
    types: list[str] = Field(default_factory=list)
    criteria: list[Criterion] = Field(default_factory=list)


class RelationshipStep(BaseModel):
    """A hop: traverse `relationship_type` (subtype-inclusive) in `direction`,
    landing on `target_types` (subtype-inclusive; empty = any). Carries NO
    criteria — filtering lives in `FilterStep`. `children` is reserved for
    post-Stage-1 branching and MUST be empty in schema v2."""

    kind: Literal["relationship"] = "relationship"
    relationship_type: str
    direction: Literal["out", "in", "either"] = "out"
    target_types: list[str] = Field(default_factory=list)
    children: list[StepItem] = Field(default_factory=list)
    #: free-form user note explaining the step's intent (UI-only; the
    #: evaluator ignores it).
    comment: str | None = None

    @model_validator(mode="after")
    def _v2_is_linear(self) -> RelationshipStep:
        if self.children:
            raise ValueError(
                "branching steps (`children`) are not supported in schema v2"
            )
        return self


class FilterStep(BaseModel):
    """Prunes the current frontier in place: keep an element iff it matches
    ALL criteria. Adds no chain column (see the evaluator). Criteria reuse the
    shared search vocabulary; property criteria are existence-gated at
    evaluation time."""

    kind: Literal["filter"] = "filter"
    criteria: list[Criterion] = Field(default_factory=list)
    #: free-form user note explaining the step's intent (UI-only; the
    #: evaluator ignores it).
    comment: str | None = None


class PropertyStep(BaseModel):
    """A hop through a property. Adds ONE chain column, like a
    RelationshipStep, with per-element behaviour decided by the element's
    EFFECTIVE property def:

    - datatype is an ELEMENT type → follow the value(s) — element ids — to the
      referenced element(s); dangling ids are skipped.
    - datatype is SCALAR → the chain TERMINATES at the property's value(s)
      (one chain per list item), carried as `PropertyValue` terminal nodes so
      the UI can display the value; no further step can extend such a chain.

    A missing def or unset value prunes the chain — graceful, mirroring
    FilterStep's existence-gating, so the engine never raises on odd models."""

    kind: Literal["property"] = "property"
    property_name: str
    #: free-form user note explaining the step's intent (UI-only; the
    #: evaluator ignores it).
    comment: str | None = None


StepItem = Annotated[
    RelationshipStep | FilterStep | PropertyStep, Field(discriminator="kind")
]


class PathNavigation(BaseModel):
    kind: Literal["path"]
    schema_version: int = SCHEMA_VERSION
    #: user-chosen display name; None keeps the UI's automatic lettering
    #: ("Path A", "Path B", ...). UI-only — the evaluator ignores it.
    name: str | None = None
    start: StartNode
    steps: list[StepItem] = Field(default_factory=list)
    #: cycle guard: when True (default, prior behavior), a chain skips any
    #: element already in its own prefix; when False, chains may revisit
    #: elements (still terminates — chain length is capped at len(steps)+1).
    exclude_visited: bool = True

    @model_validator(mode="after")
    def _cap_steps(self) -> PathNavigation:
        if len(self.steps) > MAX_STEPS:
            raise ValueError(f"a navigation may have at most {MAX_STEPS} steps")
        return self


class Operand(BaseModel):
    """One input to a set operation: a saved navigation (`ref` = artifact id)
    XOR an inline `definition`, contributing its elements at `step_index`
    (see the chain convention in the module docstring)."""

    ref: str | None = None
    definition: NavigationDefinition | None = None
    step_index: int | None = Field(default=None, ge=0)

    @model_validator(mode="after")
    def _exactly_one_source(self) -> Operand:
        if (self.ref is None) == (self.definition is None):
            raise ValueError("an operand needs exactly one of `ref` / `definition`")
        return self


class SetExpression(BaseModel):
    kind: Literal["set_op"]
    schema_version: int = SCHEMA_VERSION
    op: Literal["union", "intersection", "difference", "symmetric_difference"]
    operands: list[Operand] = Field(min_length=1)


class RowStart(BaseModel):
    """Start = the element(s) this navigation is rooted at by its caller.

    Only meaningful when `evaluate()` is given a row binding (table columns
    supply one per row). Reaching a RowStart with no binding is a ValueError:
    a row-rooted definition is not evaluable on its own."""

    kind: Literal["row"] = "row"


StartNode = Annotated[Scope | SetExpression | RowStart, Field(discriminator="kind")]
NavigationDefinition = Annotated[
    PathNavigation | SetExpression, Field(discriminator="kind")
]

RelationshipStep.model_rebuild()
FilterStep.model_rebuild()
PropertyStep.model_rebuild()
PathNavigation.model_rebuild()
Operand.model_rebuild()
SetExpression.model_rebuild()

#: validates/dumps a full definition document (artifact payloads, API bodies).
NAVIGATION_ADAPTER: TypeAdapter[NavigationDefinition] = TypeAdapter(
    NavigationDefinition
)

File: engine/src/navigation/schema.ts
/**
 * The navigation definitions of `core/navigation/schema.py`: a path walks
 * steps from a start set; a set expression combines the element sets of its
 * operands. Every chain holds its start at index 0 and one column per
 * relationship, property or script step; a filter adds none.
 */
import { ReadError } from '../read/errors.ts';
import { readCriteria, type Criterion } from '../search/criteria.ts';
import { PyFloat } from '../value/types.ts';

/** The most steps a path may hold. */
export const MAX_STEPS = 10;

/** Elements of any of `types` (subtype-inclusive; none is every type) matching every criterion. */
export type Scope = { kind: 'scope'; types: string[]; criteria: Criterion[] };

/** The element(s) the caller binds, a table row's. */
export type RowStart = { kind: 'row' };

export type RelationshipStep = {
	kind: 'relationship';
	relationship_type: string;
	direction: 'out' | 'in' | 'either';
	target_types: string[];
};

/** Keeps a chain whose last element matches every criterion; adds no column. */
export type FilterStep = { kind: 'filter'; criteria: Criterion[] };

export type PropertyStep = { kind: 'property'; property_name: string };

/** A snippet's source: neither set is an unconfigured step. The definition is not read further. */
export type SnippetSource = { ref: string | null; definition: object | null };

/** `comment` labels the step's column. */
export type ScriptStep = { kind: 'script'; snippet: SnippetSource; comment: string | null };

export type NavigationStep = RelationshipStep | FilterStep | PropertyStep | ScriptStep;

export type SetOp = 'union' | 'intersection' | 'difference' | 'symmetric_difference';

/** A saved navigation (`ref`) or an inline one, contributing its elements at `step_index` (none: the last). */
export type Operand = {
	ref: string | null;
	definition: NavigationDefinition | null;
	step_index: number | null;
};

export type SetExpression = { kind: 'set_op'; op: SetOp; operands: Operand[] };

export type PathNavigation = {
	kind: 'path';
	start: Scope | SetExpression | RowStart;
	steps: NavigationStep[];
	exclude_visited: boolean;
};

export type NavigationDefinition = PathNavigation | SetExpression;

// -- reading -------------------------------------------------------------------

type Doc = { readonly [key: string]: unknown };

const isDoc = (value: unknown): value is Doc =>
	typeof value === 'object' &&
	value !== null &&
	!Array.isArray(value) &&
	!(value instanceof PyFloat);

function refuse(where: string, message: string): never {
	throw new ReadError(422, `${where}: ${message}`);
}

// An absent key takes the default; a `null` does not, as pydantic has it.
const field = (doc: Doc, key: string, fallback?: unknown): unknown =>
	Object.hasOwn(doc, key) && doc[key] !== undefined ? doc[key] : fallback;

function doc(raw: unknown, where: string): Doc {
	if (!isDoc(raw)) refuse(where, 'must be an object');
	return raw;
}

function str(d: Doc, key: string, where: string): string {
	const value = field(d, key);
	if (typeof value !== 'string') refuse(`${where}.${key}`, 'must be a string');
	return value;
}

function optionalStr(d: Doc, key: string, where: string): string | null {
	const value = field(d, key, null);
	if (value !== null && typeof value !== 'string')
		refuse(`${where}.${key}`, 'must be a string or null');
	return value;
}

function oneOf<T extends string>(
	d: Doc,
	key: string,
	where: string,
	options: readonly T[],
	fallback?: T
): T {
	const value = field(d, key, fallback);
	if (!options.includes(value as T)) {
		refuse(`${where}.${key}`, `must be one of ${options.join(', ')}`);
	}
	return value as T;
}

function names(d: Doc, key: string, where: string): string[] {
	const value = field(d, key, []);
	if (!Array.isArray(value) || !value.every((name) => typeof name === 'string')) {
		refuse(`${where}.${key}`, 'must be a list of strings');
	}
	return [...(value as string[])];
}

function list(d: Doc, key: string, where: string): readonly unknown[] {
	const value = field(d, key, []);
	if (!Array.isArray(value)) refuse(`${where}.${key}`, 'must be a list');
	return value;
}

/** Checked, not kept: an int when present. */
function checkSchemaVersion(d: Doc, where: string): void {
	const value = field(d, 'schema_version', 0);
	if (!((typeof value === 'number' && Number.isInteger(value)) || typeof value === 'bigint')) {
		refuse(`${where}.schema_version`, 'must be an integer');
	}
}

const criteriaOf = (d: Doc, where: string): Criterion[] =>
	readCriteria(field(d, 'criteria', []), `${where}.criteria`);

function readStep(raw: unknown, where: string): NavigationStep {
	const d = doc(raw, where);
	const kind = oneOf(d, 'kind', where, ['relationship', 'filter', 'property', 'script'] as const);
	// Every step's comment is checked; only a script step's labels its column.
	const comment = optionalStr(d, 'comment', where);
	switch (kind) {
		case 'relationship': {
			const step: RelationshipStep = {
				kind,
				relationship_type: str(d, 'relationship_type', where),
				direction: oneOf(d, 'direction', where, ['out', 'in', 'either'] as const, 'out'),
				target_types: names(d, 'target_types', where)
			};
			if (list(d, 'children', where).length > 0) {
				refuse(`${where}.children`, 'branching steps (`children`) are not supported in schema v2');
			}
			return step;
		}
		case 'filter':
			return { kind, criteria: criteriaOf(d, where) };
		case 'property':
			return { kind, property_name: str(d, 'property_name', where) };
		case 'script': {
			const at = `${where}.snippet`;
			const snippet = doc(field(d, 'snippet', {}), at);
			const ref = optionalStr(snippet, 'ref', at);
			const definition = field(snippet, 'definition', null);
			if (definition !== null && !isDoc(definition)) {
				refuse(`${at}.definition`, 'must be an object or null');
			}
			if (ref !== null && definition !== null) {
				refuse(at, 'provide at most one of `ref` / `definition`');
			}
			return { kind, snippet: { ref, definition }, comment };
		}
	}
}

function readOperand(raw: unknown, where: string): Operand {
	const d = doc(raw, where);
	const ref = optionalStr(d, 'ref', where);
	const inline = field(d, 'definition', null);
	const definition = inline === null ? null : readDefinition(inline, `${where}.definition`);
	const stepIndex = field(d, 'step_index', null);
	if (
		stepIndex !== null &&
		!(typeof stepIndex === 'number' && Number.isInteger(stepIndex) && stepIndex >= 0)
	) {
		refuse(`${where}.step_index`, 'must be an integer of at least 0 or null');
	}
	if ((ref === null) === (definition === null)) {
		refuse(where, 'an operand needs exactly one of `ref` / `definition`');
	}
	return { ref, definition, step_index: stepIndex };
}

const SET_OPS: readonly SetOp[] = ['union', 'intersection', 'difference', 'symmetric_difference'];

function readSet(d: Doc, where: string): SetExpression {
	checkSchemaVersion(d, where);
	const op = oneOf(d, 'op', where, SET_OPS);
	const raw = field(d, 'operands');
	if (!Array.isArray(raw)) refuse(`${where}.operands`, 'must be a list');
	if (raw.length === 0) refuse(`${where}.operands`, 'must hold at least one operand');
	return {
		kind: 'set_op',
		op,
		operands: raw.map((operand: unknown, i) => readOperand(operand, `${where}.operands[${i}]`))
	};
}

function readStart(raw: unknown, where: string): PathNavigation['start'] {
	const d = doc(raw, where);
	const kind = oneOf(d, 'kind', where, ['scope', 'set_op', 'row'] as const);
	if (kind === 'row') return { kind };
	if (kind === 'set_op') return readSet(d, where);
	return { kind, types: names(d, 'types', where), criteria: criteriaOf(d, where) };
}

function readDefinition(raw: unknown, where: string): NavigationDefinition {
	const d = doc(raw, where);
	const kind = oneOf(d, 'kind', where, ['path', 'set_op'] as const);
	if (kind === 'set_op') return readSet(d, where);
	checkSchemaVersion(d, where);
	optionalStr(d, 'name', where);
	const start = readStart(field(d, 'start'), `${where}.start`);
	const steps = list(d, 'steps', where).map((step, i) => readStep(step, `${where}.steps[${i}]`));
	if (steps.length > MAX_STEPS) {
		refuse(`${where}.steps`, `a navigation may have at most ${MAX_STEPS} steps`);
	}
	const excludeVisited = field(d, 'exclude_visited', true);
	if (typeof excludeVisited !== 'boolean') refuse(`${where}.exclude_visited`, 'must be a boolean');
	return { kind, start, steps, exclude_visited: excludeVisited };
}

/**
 * A definition as a client sends it or an artifact holds it, read as pydantic
 * reads it in canonical JSON: every `kind`, `op` and criterion `type`
 * required, the defaults filled, unknown keys ignored. `schema_version`, a
 * path's `name` and a step's `comment` are checked, never kept, bar a script
 * step's comment. What pydantic would
 * coerce is refused, in the engine's words where the core has none.
 */
export function readNavigation(raw: unknown, path: string): NavigationDefinition {
	return readDefinition(raw, path);
}
